### PR TITLE
[3.1] Use native code slot for default interface methods (#25770)

### PR DESCRIPTION
**Pull request by kouvel**
- Partial port of https://github.com/dotnet/coreclr/pull/25770 to 3.1 (only the main change, excludes unrelated cleanup)
- Use native code slot for default interface methods so that `MethodDesc::GetNativeCode()` can retrieve the current native code entry point (instead of returning null as before), and code versioning can find a matching code version from the code start address
- Interface methods currently require having a precode, so the "method entry point" can't be used to directly store the native code entry point
- Reenabled a couple of default interface method tests under GCStress

Fixes https://github.com/dotnet/coreclr/issues/25690

### Customer impact

Mostly only affects some GCStress tests. The change can have harmless non-test effects, for example a default interface method may be jitted redundantly fewer times after the change (as expected), since before the change the call to `GetNativeCode()` here would always return null for a default interface method and the method may be jitted again unnecessarily.
https://github.com/dotnet/coreclr/blob/867aa5e1141c8801796a80171a3749d944c6b9af/src/vm/codeversion.cpp#L2304-L2309

Otherwise, the main purpose of the port is to fix some GCStress tests to make CI runs more reliable and useful for testing other fixes made to 3.1.

### Testing

@echesakovMSFT has verified that all related GCStress failures are fixed after the change on Windows arm64 where the issues were seen.

### Risk

Low:
- The change has been in all of the 5.0 releases
- The change makes tracking of the native code for default interface methods similar to other methods that have a precode and don't have a stable entry point, so the behavior would be more consistent with other types of methods
- The native code slot for default interface methods is a pointer-size extra memory per such method, it wouldn't be significant as the number of such methods is typically low compared to other types of methods, and most other methods already use a native code slot

### CODE REVIEW

**Review by MichalStrehovsky — APPROVED** [review]

**Comment by aik-jahoda** [comment]
@kouvel,  @MichalStrehovsky, do we need servise approval? This PR looks like product change

**Comment by kouvel** [comment]
Yes it needs to go through shiproom, should wait for @noahfalk's review as well

**Review by noahfalk — APPROVED** [review]
The change seems OK but I'll caveat that I have no significant knowledge of implementation details of Default Interface Methods. If there is some corner case for how default interface methods interact relative to other methods its unlikely I would have caught it. I'm relying on this behavior seeming reasonable for other methods + has decent test coverage + shipped in .NET 5.

**Comment by kouvel** [comment]
Thanks all. I'm also not an expert on the implementation details of default interface methods. @jkotas, would you be able to take a quick look as well?

**Comment by jkotas** [comment]
> Mostly only affects some GCStress tests.

This does not sound like a good enough justification for pushing this change into 3.1 servicing.

**Comment by kouvel** [comment]
In non-test code paths that would be affected by the change I didn't find any significant ones that would add higher risk, but I understand that it may not be worth the risk. @echesakovMSFT / @BruceForstall, would it be reasonable to disable the affected GCStress tests in 3.1 instead?

**Review by jeffschwMSFT — APPROVED** [review]
Approved.  I will take for consideration in 3.1.x

**Comment by jeffschwMSFT** [comment]
@Anipik would this make the cut off for merging in the inflight servicing release for April?  Thanks

**Comment by Anipik** [comment]
> @Anipik would this make the cut off for merging in the inflight servicing release for April? Thank

yes we can merge this for april release

**Pull request merged**
